src/Microsoft.DotNet.Interactive.CSharp/CSharpKernel.cs

```diff
@@ -477,5 +491,33 @@ void ISupportNuget.RegisterResolvedPackageReferences(IReadOnlyList<ResolvedPacka
 
         public IEnumerable<string> RestoreSources =>
             PackageRestoreContext.RestoreSources;
+
+        public Task HandleAsync(RequestValueNames command, KernelInvocationContext context)
+        {
+            context.Publish(new ValueNamesProduced(GetVariableNames(), command));
+            return Task.CompletedTask;
+        }
+
+        public Task HandleAsync(RequestValue command, KernelInvocationContext context)
+        {
+            if (TryGetVariable(command.Name, out var value))
+            {
+                var formattedValues = new List<FormattedValue>();
+                if (command.MimeTypes?.Any() == true)
+                {
+                    formattedValues.AddRange(command.MimeTypes.Select(mimeType => new FormattedValue(mimeType, value?.ToDisplayString(mimeType))));
+                }
+                else
+                {
+                    var preferredMimeType = Formatter.GetPreferredMimeTypeFor(value?.GetType() ?? typeof(object));
+                    formattedValues.Add(new FormattedValue(preferredMimeType, value?.ToDisplayString(preferredMimeType)));
+                }
+
+                context.Publish(new ValueProduced(value, command.Name, command, formattedValues));
+                return Task.CompletedTask;
+            }
+
+            throw new ValueNotFoundException(command.Name);
```

Comment:
Let's avoid introducing a custom exception type for this, especially as it's going to be wrapped into a `CommandFailed` anyway.

Maybe an `InvalidOperationException`?

Reply:
ok